nbbuild/misc/prepare-bundles/src/main/java/org/netbeans/prepare/bundles/PrepareBundles.java

```diff
@@ -0,0 +1,295 @@
+        Path targetDir = Paths.get(args[0]);
+        Path packagesDir = targetDir.resolve("package");
+        new ProcessBuilder("npm", "install").directory(packagesDir.toFile()).inheritIO().start().waitFor();
+        Path bundlesDir = targetDir.resolve("bundles");
+        Files.createDirectories(bundlesDir);
+        try (DirectoryStream<Path> ds = Files.newDirectoryStream(bundlesDir)) {
+            for (Path bundle : ds) {
+                Files.delete(bundle);
+            }
+        }
+
+        Path licensesDir = targetDir.resolve("licenses");
+
+        Files.createDirectories(licensesDir);
+        try (DirectoryStream<Path> ds = Files.newDirectoryStream(licensesDir)) {
+            for (Path license : ds) {
+                Files.delete(license);
+            }
+        }
+
+        Path externalDir = targetDir.resolve("external");
+        Files.createDirectories(externalDir);
+        try (DirectoryStream<Path> ds = Files.newDirectoryStream(externalDir)) {
+            for (Path external : ds) {
+                Files.delete(external);
+            }
+        }
+
+        Map<List<String>, LicenseUses> tokens2Projects = new HashMap<>();
+        Map<String, LicenseDescription> project2License = new HashMap<>();
+        Map<String, String> project2Notice = new HashMap<>();
+        try (DirectoryStream<Path> ds = Files.newDirectoryStream(packagesDir.resolve("node_modules"));
+             Writer binariesList = new OutputStreamWriter(Files.newOutputStream(bundlesDir.resolve("binaries-list")), "UTF-8")) {
+            for (Path module : ds) {
+                if (".bin".equals(module.getFileName().toString())) continue;
+                if ("@types".equals(module.getFileName().toString())) continue;
+                Path packageJson = module.resolve("package.json");
+
+                if (!Files.isReadable(packageJson)) {
+                    throw new IllegalStateException("Cannot find package.json for: " + module.getFileName());
+                }
+
+                String packageJsonText = readString(packageJson);
+                Map<String, Object> packageJsonData = new Gson().fromJson(packageJsonText, HashMap.class);
+                String name = (String) packageJsonData.get("name");
+                String version = (String) packageJsonData.get("version");
+                String description = (String) packageJsonData.get("description");
+                String homepage = (String) packageJsonData.get("homepage");
+                String licenseKey = (String) packageJsonData.get("license");
+
+                String licenseText = null;
+
+                for (String l : LICENSE_FILE_NAMES) {
+                    if (Files.isReadable(module.resolve(l))) {
+                        licenseText = readString(module.resolve(l));
+                        break;
+                    }
+                }
+
+                if (licenseText == null) {
+                    String hardcodedLicenseName = name + "-" + version + "-license";
```

Comment:
I understand that this allows supplying license information for packages, that can't be correctly parsed. Looks ok, though I see potential problems in the future, if we get into the habit of bundling "many" JS libraries.

I would consider removing the null check in 132. That way invalid license info could be overriden.

Reply:
Yes, the more libraries we handle, the more difficult this gets - but still hopefully easier than doing all this by hand. (Esp. the VS Code extensions uses a lot of libraries - much more than I'd expect.)